Add more filters to learners screen

raq929 opened the pull request:
This adds a course title and a date filter to the admin portal. 

MIRO Board: https://miro.com/app/board/o9J_kp-7aks=/?moveToWidget=3074457348864300777&cot=13

**Notes for review**
The backend has been updated but not merged, so you will need to be on the `ENT-3086` filters branch of `edx-enterprise-data` (installed into `edx-analytics-data-api`, similarly to how we installed `edx-enterprise` into `edx-platform`) in order to review this properly.

I created courses and learners by using the EnterpriseEnrollmentFactory in the django shell.

The date field is disabled until a course title has been selected.

There are plans for a form reset button and some help text in a pop-over, but I'm going to open another ticket for those because the design has not been finalized.

binodpant commented:
@raq929 I would love to try this locally but I have never setup edx-enterprise-data and edx-analytics-data-api before. 

When I naively tried to guess and run the management command in the edx-enterprise-data repo I got this, I used a generated uuid from some web tool:

```
(venv3.5.6) ~/work/edx-enterprise-data (ENT-3086-filtering ✔) ᐅ ./manage.py create_enterprise_user 6ad30ad9-c5dc-4343-825d-08036d06045f
CommandError: Error trying to create EnterpriseUser with uuid 6ad30ad9-c5dc-4343-825d-08036d06045f: no such table: enterprise_user
```

Prior to this I have done this:

```
~/work/edx-analytics-data-api (master ✔) ᐅ ./manage.py set_api_key test-user-for-ent-3086 1234
Set API key for user test-user-for-ent-3086 to 1234
```

and this

```
(venv3.5.6) ~/work/edx-analytics-data-api (master ✔) ᐅ make loaddata
./manage.py migrate --noinput --run-syncdb --database=default
```

But I presume the above are not sufficient

raq929 commented:
Here's what I did. I may be missing steps, so sorry if that's the case.
1. moved `edx-analytics-data-api` into it's own folder and created a virtual env for it.
1. moved `edx-enterprise-data` into an `src` directory in the same directory as `edx-analytics-data-api`
1. `cd {{folder}}/edx-analytics-data-api`, activate the virtualenv
1. `make requirements` on in the virtualenv for `edx-analytics-data-api`
1.  ran `pip install -e ./src/edx-enterprise-data`
1. ...
1. PROFIT

raq929 commented:
In the frontend app:
1. `npm install`
1. `npm start`
1. Login, get redirected to localhost:18000
1. Manually reenter localhost:1991
1. Think about how many more times you will do that before you get annoyed enough to change it.

binodpant commented:
My devstack is totally broken right now so I will do my best to just code review this tomorrow morning...sorry gotta leave for now

binodpant commented:
Thanks, so after (somehow) fixing up my devstack, I followed above steps and when I now load the page and log in as 'edx' I get

on frontend: 403

on backend of edx-analytics-data-api:

```
(venv3.5.6) ~/work/edx-analytics-data-api (master ✔) ᐅ ./manage.py runserver
Performing system checks...

System check identified no issues (0 silenced).
July 28, 2020 - 02:01:29
Django version 2.2.14, using settings 'analyticsdataserver.settings.local'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
Testing (request_user_has_implicit_access | request_user_has_explicit_access)
Testing (request_user_has_implicit_access | request_user_has_explicit_access)
  request_user_has_implicit_access = False
  request_user_has_implicit_access = False
  request_user_has_explicit_access = False
  (request_user_has_implicit_access | request_user_has_explicit_access) = False
  request_user_has_explicit_access = False
  (request_user_has_implicit_access | request_user_has_explicit_access) = False
```

raq929 commented:
I forgot a step! In the enterprise admin portal
localhost:18000/admin/enterprise, in the same place you enabled the learner portal, you can enable this admin portal. I'm gonna open a ticket to have both of those turned on by default when the test enterprise is created...

binodpant commented on src/components/Admin/AdminSearchForm.jsx:
```diff
@@ -0,0 +1,163 @@
+    const queryParams = qs.parse(location.search);
+    this.state = {
+      searchQuery: queryParams.search,
+      searchCourseQuery: queryParams.search_course,
```
Ah I see this snake_casing must be from python right? In js we typically do cameCasing, and I thought in the other FE libs we use a function to convert from pythonic to js-ic ... would that apply here too? would that be too much work? IT would be nice in my opinion, but I would not fret much if that adds too much work... just a thought

binodpant commented on src/components/Admin/AdminSearchForm.jsx:
```diff
@@ -0,0 +1,163 @@
+      const { search, search_course, search_start_date } = qs.parse(location.search);
+      const {
+        search: prevSearch,
+        search_course: prevSearchCourse,
```
this is just convention/style wise: hmm to be consistent, could these also be camelCased? like searchCourse? or would that complicate work in sending over to the backend? I think some FE libs use a conversion pattern from snake to camel and back .. wondering if we want that here too

raq929 replied:
Yeah, I thought about that. I'm parsing from the query string, so... I don't know. I don't think I want that here, I want it to be clear what's coming from the querystring.

binodpant commented on src/components/Admin/AdminSearchForm.jsx:
```diff
@@ -0,0 +1,163 @@
+
+    return (
+      <div className="row">
+        <div className="col-12 pr-md-0 mb-0">
```
wow the styling here seems to have gotten interesting. I am still trying to digest the various col > row gridness here ... maybe better once I have this UI locally

raq929 replied:
It makes it look nice on mobile. 😉

binodpant commented on src/components/Admin/AdminSearchForm.jsx:
```diff
@@ -0,0 +1,163 @@
+              <Form inline role="search">
+                <div className="col-6 pl-0 pr-md-2 pr-lg-3">
+                  <Form.Group>
+                    <Form.Label id="course-title-search" className="search-label mb-2">Filter by courses</Form.Label>
```
Nit: As per the miro spec it says 'Filter by course'

binodpant commented on src/components/Admin/AdminSearchForm.jsx:
```diff
@@ -0,0 +1,163 @@
+                </div>
+                <div className="col-6 pr-0 px-md-2 px-lg-3">
+                  <Form.Group>
+                    <Form.Label id="date-search" className="search-label mb-2">Filter by start date</Form.Label>
```
Did we decide to not do the help icon as per the miro spec, next to this label?

raq929 replied:
We didn't have the text yet, I opened a separate ticket for it: https://openedx.atlassian.net/browse/ENT-3235

binodpant commented on src/components/Admin/index.jsx:
```diff
@@ -289,43 +259,15 @@ class Admin extends React.Component {
-      lastUpdatedDate,
       match,
     } = this.props;
+    // const {searchQuery, searchCourseQuery, searchDateQuery} = this.state;
```
remove this line?

raq929 replied:
oops, yes, thanks.

binodpant commented on src/components/Admin/index.jsx:
```diff
@@ -365,16 +307,28 @@ class Admin extends React.Component {
+                          <div className="col-12 col-md-6  col-xl-4 pt-1 pb-3">
+                            {lastUpdatedDate &&
+                              <React.Fragment>
+                                Showing data as of {formatTimestamp({ timestamp: lastUpdatedDate })}
```
I can't see this text on the ui.. how would I test this piece?

raq929 replied:
Honestly, I don't know what makes `lastUpdatedDate` show up. (It's not my original code, just moved tot this location.)
For UI testing I just changed the `lastUpdatedDate` before the and to `true`.

binodpant commented:
Functionally I tested everything appears to work locally, tried using the filter for course and date. Also verified the urls are valid when loaded in a new tab, after selecting filters, all filters get loaded

just have some thoughts/comments on minor stuff above but once we resolve those conversations, no other major flags! Looking good! The dropdown style will get autofixed I guess by someone as referred to in dev. 

Also ran the Axe tool in chrome, to check accessbility and it did not flag anything for the area worked on for this PR so great!
